sql-plugin/src/main/scala/com/nvidia/spark/rapids/literals.scala

```diff
@@ -79,7 +91,132 @@ object GpuScalar {
+  }
+
+  private def convertDecimalTo(dec: Decimal, dt: DecimalType): JLong = dec match {
+    case null => null
+    case d if d.precision > dt.precision =>
+      throw new IllegalArgumentException(s"Decimal $d exceeds precision constraint of $dt")
+    case _ => dec.toUnscaledLong
+  }
+
+  /** Converts an element for nested lists */
+  private def convertElementTo(element: Any, elementType: DataType): Any = elementType match {
+    case StringType => convertUTF8StringTo(element.asInstanceOf[UTF8String])
+    case dt: DecimalType =>
+      if (DecimalType.is32BitDecimalType(dt)) {
+        convertDecimalTo(element.asInstanceOf[Decimal], dt).intValue()
+      } else {
+        convertDecimalTo(element.asInstanceOf[Decimal], dt)
+      }
+    case ArrayType(eType, _) => element.asInstanceOf[ArrayData] match {
+      case null => null
+      case ar => ar.array.map(convertElementTo(_, eType)).toList.asJava
+    }
+    case StructType(fields) => element.asInstanceOf[InternalRow] match {
+      case null => null
+      case row =>
+        val data = fields.zipWithIndex.map { case (f, id) =>
+          convertElementTo(row.get(id, f.dataType), f.dataType)
+        }
+        new HostColumnVector.StructData(data.asInstanceOf[Array[Object]]: _*)
+    }
+    case _ => element
+  }
+
+  /**
+   * Creates a cudf Scalar from a ArrayData.
+   *
+   * It does not handle null because null is filtered out in 'from'.
+   *
+   * NOTE: This is done by leveraging the `ColumnVector.fromXXX` API family. These APIs are
+   * designed for test only according to its doc, but it is ok to use here, since the size
+   * of the literal values will not be large in most cases.
+   *
+   * @param data the data to build the Scalar. Should not be null.
+   * @param elementType the element type
+   * @return a cudf Scalar contains the array data.
+   */
+  private def createScalarFromArray(data: ArrayData, elementType: DataType): Scalar = {
+    val array = data.array
+    val listView = elementType match {
+      // Uses the boxed version for primitive types to keep the nulls.
+      case ByteType => ColumnVector.fromBoxedBytes(castTo[JByte](array): _*)
+      case DateType => ColumnVector.timestampDaysFromBoxedInts(castTo[Integer](array): _*)
+      case LongType => ColumnVector.fromBoxedLongs(castTo[JLong](array): _*)
+      case ShortType => ColumnVector.fromBoxedShorts(castTo[JShort](array): _*)
+      case FloatType => ColumnVector.fromBoxedFloats(castTo[JFloat](array): _*)
+      case DoubleType => ColumnVector.fromBoxedDoubles(castTo[JDouble](array): _*)
+      case IntegerType => ColumnVector.fromBoxedInts(castTo[Integer](array): _*)
+      case BooleanType => ColumnVector.fromBoxedBooleans(castTo[JBoolean](array): _*)
+      case TimestampType =>
+        ColumnVector.timestampMicroSecondsFromBoxedLongs(castTo[JLong](array): _*)
+      case StringType =>
+        ColumnVector.fromStrings(castTo[UTF8String, String](array)(convertUTF8StringTo(_)): _*)
```

Comment:
This copies the string _three times_: once when converting from UTF8String to String, another when getting the bytes out of the string, and yet another when copying the bytes to the device.  It would be nice to completely avoid Java's `String` type when dealing with Spark's `StringType` since `String` is so expensive to wield due to the required copying.  Using `HostColumnVector.Builder`'s `appendUTF8String` would be more efficient here.  Would be nice if `ColumnVector` had a `fromUTF8Strings` that took a vararg of `byte[]` so we could avoid the `HostColumnVector.Builder` direct usage.

Reply:
Good to know, updated to use `ColumnVector.build` now.
The followup issue is here https://github.com/rapidsai/cudf/issues/8137